op-service/dial/active_l2_provider_test.go

```diff
@@ -0,0 +1,195 @@
+	primarySequencer.ExpectClose()
+	secondarySequencer := new(testutils.MockRollupClient)
+	secondarySequencer.ExpectSequencerActive(true, nil)
+
+	mockRollupDialer := func(ctx context.Context, timeout time.Duration, log log.Logger, url string) (RollupClientInterface, error) {
+		if url == "primaryRollup" {
+			return primarySequencer, nil
+		} else if url == "secondaryRollup" {
+			return secondarySequencer, nil
+		} else {
+			return nil, fmt.Errorf("unknown test url: %s", url)
+		}
+	}
+
+	endpointProvider, err := newActiveL2RollupProvider(
+		context.Background(),
+		[]string{"primaryRollup", "secondaryRollup"},
+		1*time.Microsecond,
+		1*time.Minute,
+		testlog.Logger(t, log.LvlDebug),
+		mockRollupDialer,
+	)
+	require.NoError(t, err)
+	// Check that the first client is used, then the second once the first declares itself inactive.
+	firstSequencerUsed, err := endpointProvider.RollupClient(context.Background())
+	require.NoError(t, err)
+	require.Same(t, primarySequencer, firstSequencerUsed)
+	secondSequencerUsed, err := endpointProvider.RollupClient(context.Background())
+	require.NoError(t, err)
+	require.Same(t, secondarySequencer, secondSequencerUsed)
+}
+
+// TestActiveSequencerFailoverBehavior_L2Providers_Inactive tests that the ActiveL2EndpointProvider
+// will failover to the next provider if the current one is not active.
+func TestActiveSequencerFailoverBehavior_L2Providers_Inactive(t *testing.T) {
+	// as TestActiveSequencerFailoverBehavior_RollupProviders,
+	// but ensure the added `EthClient()` method also triggers the failover.
+	primarySequencer := new(testutils.MockRollupClient)
+	primarySequencer.ExpectSequencerActive(true, nil)
+	primarySequencer.ExpectSequencerActive(false, nil)
+	primarySequencer.ExpectClose()
+	secondarySequencer := new(testutils.MockRollupClient)
+	secondarySequencer.ExpectSequencerActive(true, nil)
+
+	mockRollupDialer := func(ctx context.Context, timeout time.Duration, log log.Logger, url string) (RollupClientInterface, error) {
+		if url == "primaryRollup" {
+			return primarySequencer, nil
+		} else if url == "secondaryRollup" {
+			return secondarySequencer, nil
+		} else {
+			return nil, fmt.Errorf("unknown test url: %s", url)
+		}
+	}
+	primaryEthClient := new(testutils.MockEthClient)
+	primaryEthClient.ExpectClose()
+	secondaryEthClient := new(testutils.MockEthClient)
+	mockEthDialer := func(ctx context.Context, timeout time.Duration, log log.Logger, url string) (EthClientInterface, error) {
+		if url == "primaryEth" {
+			return primaryEthClient, nil
+		} else if url == "secondaryEth" {
+			return secondaryEthClient, nil
+		} else {
+			return nil, fmt.Errorf("unknown test url: %s", url)
+		}
+	}
+	endpointProvider, err := newActiveL2EndpointProvider(
+		context.Background(),
+		[]string{"primaryEth", "secondaryEth"},
+		[]string{"primaryRollup", "secondaryRollup"},
+		1*time.Microsecond,
+		1*time.Minute,
+		testlog.Logger(t, log.LvlDebug),
+		mockEthDialer,
+		mockRollupDialer,
+	)
```

Comment:
A lot of the setup code could is duplicated in tests. I think adding a test harness could be beneficial here and make it easier to write more tests, while make the existing ones more readable, like
```go
type endpointProviderTest struct {
    rollupClients []*testutils.MockRollupClient
    ethClients []*testutils.MockEthClient
    // dialers, ...
}

func setupEndpointProviderTest(t * testing.T, numSequencers int) *endpointProviderTest {
    // all the setup, except setting up the mocks
}
```

to make it possible to setup the mocks before calling the constructor, you could add a method to the harness like
```go
func (et *endpointProviderTest) newActiveL2EndpointProvider() (*ActiveL2EndpointProvider, error)
```